File: course-schedule.py

class Solution:
    def canFinish(self, num_courses, prerequisites):
        graph = {i: [] for i in range(num_courses)}

        for to_take, need_to in prerequisites:
            graph[to_take].append(need_to)
            
        visited = [False] * num_courses
        visited_stack = [False] * num_courses
        
        def dfs(node):
            visited[node] = True
            visited_stack[node] = True
            
            for child in graph[node]:
                if visited_stack[child]: return False
                if not visited[child]:
                    if not dfs(child):
                        return False
            visited_stack[node] = False
            return True
            
        for node in graph:
            if not visited[node]:
                if not dfs(node):
                    return False
                
        return True

assert Solution().canFinish(1, []) == True
assert Solution().canFinish(2, [[0,1]]) == True
assert Solution().canFinish(2, [[1,0]]) == True
assert Solution().canFinish(2, [[1,0],[0,1]]) == False
assert Solution().canFinish(3, [[0,1],[0,2],[1,2]]) == True

File: java/myjava/CourseSchedule.java
package myjava;

import java.util.*;
import java.util.stream.*;

/*-
https://leetcode.com/problems/course-schedule/description/

There are a total of numCourses courses you have to take, labeled from 0 to numCourses - 1. You are given an
array prerequisites where prerequisites[i] = [ai, bi] indicates that you must take course bi first if you want
to take course ai.

For example, the pair [0, 1], indicates that to take course 0 you have to first take course 1.
Return true if you can finish all courses. Otherwise, return false.

Example 1:
Input: numCourses = 2, prerequisites = [[1,0]]
Output: true
Explanation: There are a total of 2 courses to take. 
To take course 1 you should have finished course 0. So it is possible.

Example 2:
Input: numCourses = 2, prerequisites = [[1,0],[0,1]]
Output: false
Explanation: There are a total of 2 courses to take. 
To take course 1 you should have finished course 0, and to take course 0 you should also have finished course 1.
So it is impossible.

Constraints:

1 <= numCourses <= 2000
0 <= prerequisites.length <= 5000
prerequisites[i].length == 2
0 <= ai, bi < numCourses
All the pairs prerequisites[i] are unique.
*/

public class CourseSchedule {
  private static boolean dfs(
    Map<Integer, Set<Integer>> graph,
    boolean[] globalVisited,
    boolean[] visited,
    Integer node
  ) {
    if (globalVisited[node]) return true;
    if (visited[node]) return false;
    visited[node] = true;
    for (Integer child : graph.get(node)) {
      if (dfs(graph, globalVisited, visited, child) == false) return false;
    }
    globalVisited[node] = true;
    return true;
  }

  public static boolean canFinish(int numCourses, int[][] prerequisites) {
    Map<Integer, Set<Integer>> graph = new HashMap<>();
    for (var prerequisite : prerequisites) {
      graph.computeIfAbsent(prerequisite[0], dependent -> new HashSet<>()).add(prerequisite[1]);
    }
    for (int i = 0; i < numCourses; i++) graph.putIfAbsent(i, new HashSet<>());

    boolean[] globalVisited = new boolean[graph.size()];
    for (Integer node : graph.keySet()) {
      if (dfs(graph, globalVisited, new boolean[graph.size()], node) == false)
        return false;
    }
    return true;
  }

  public static void main(String[] args) {
    assert canFinish(2, new int[][]{{1, 0}}) == true;
    assert canFinish(2, new int[][]{{1, 0}, {0, 1}}) == false;
  }
}
